// @dydxprotocol/exchange-wrappers/dist/src/exchange_wrappers/OasisV3MatchingExchangeWrapper.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var web3_utils_1 = __importDefault(require("web3-utils"));
var bignumber_js_1 = __importDefault(require("bignumber.js"));
var deployed_json_1 = require("../../migrations/deployed.json");
var OasisV3MatchingExchangeWrapper = /** @class */ (function () {
    function OasisV3MatchingExchangeWrapper(networkId) {
        this.networkId = networkId;
    }
    OasisV3MatchingExchangeWrapper.prototype.setNetworkId = function (networkId) {
        this.networkId = networkId;
    };
    OasisV3MatchingExchangeWrapper.prototype.getAddress = function () {
        return deployed_json_1.OasisV3MatchingExchangeWrapper[this.networkId.toString()].address;
    };
    OasisV3MatchingExchangeWrapper.prototype.orderToBytes = function (order) {
        if (!order.maxPrice) {
            return [];
        }
        var price = new bignumber_js_1.default(order.maxPrice);
        var priceDenominator = new bignumber_js_1.default('1e18');
        var priceNumerator = price.times(priceDenominator).integerValue(bignumber_js_1.default.ROUND_DOWN);
        return []
            .concat(this.toBytes(priceNumerator))
            .concat(this.toBytes(priceDenominator));
    };
    OasisV3MatchingExchangeWrapper.prototype.toBytes = function (val) {
        return web3_utils_1.default.hexToBytes(web3_utils_1.default.padLeft(web3_utils_1.default.toHex(val), 64, '0'));
    };
    return OasisV3MatchingExchangeWrapper;
}());
exports.OasisV3MatchingExchangeWrapper = OasisV3MatchingExchangeWrapper;
