// @dydxprotocol/exchange-wrappers/dist/src/exchange_wrappers/OasisV2SimpleExchangeWrapper.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var web3_utils_1 = __importDefault(require("web3-utils"));
var deployed_json_1 = require("../../migrations/deployed.json");
var OasisV2SimpleExchangeWrapper = /** @class */ (function () {
    function OasisV2SimpleExchangeWrapper(networkId) {
        this.networkId = networkId;
    }
    OasisV2SimpleExchangeWrapper.prototype.setNetworkId = function (networkId) {
        this.networkId = networkId;
    };
    OasisV2SimpleExchangeWrapper.prototype.getAddress = function () {
        return deployed_json_1.OasisV2SimpleExchangeWrapper[this.networkId.toString()].address;
    };
    OasisV2SimpleExchangeWrapper.prototype.orderToBytes = function (order) {
        return this.toBytes(order.id);
    };
    OasisV2SimpleExchangeWrapper.prototype.toBytes = function (val) {
        return web3_utils_1.default.hexToBytes(web3_utils_1.default.padLeft(web3_utils_1.default.toHex(val), 64, '0'));
    };
    return OasisV2SimpleExchangeWrapper;
}());
exports.OasisV2SimpleExchangeWrapper = OasisV2SimpleExchangeWrapper;
