// @dydxprotocol/exchange-wrappers/dist/src/exchange_wrappers/TestExchangeWrapper.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var web3_utils_1 = __importDefault(require("web3-utils"));
var deployed_json_1 = require("../../migrations/deployed.json");
var TestExchangeWrapper = /** @class */ (function () {
    function TestExchangeWrapper(networkId) {
        this.networkId = networkId;
    }
    TestExchangeWrapper.prototype.getAddress = function () {
        return deployed_json_1.TestExchangeWrapper[this.networkId.toString()].address;
    };
    TestExchangeWrapper.prototype.setNetworkId = function (networkId) {
        this.networkId = networkId;
    };
    TestExchangeWrapper.prototype.orderToBytes = function (order) {
        return []
            .concat(this.toBytes(order.originator))
            .concat(this.toBytes(order.makerToken))
            .concat(this.toBytes(order.takerToken))
            .concat(this.toBytes(order.makerAmount))
            .concat(this.toBytes(order.takerAmount))
            .concat(this.toBytes(order.allegedTakerAmount))
            .concat(this.toBytes(order.desiredMakerAmount));
    };
    TestExchangeWrapper.prototype.toBytes = function (val) {
        return web3_utils_1.default.hexToBytes(web3_utils_1.default.padLeft(web3_utils_1.default.toHex(val), 64, '0'));
    };
    return TestExchangeWrapper;
}());
exports.TestExchangeWrapper = TestExchangeWrapper;
