// @dydxprotocol/exchange-wrappers/dist/src/exchange_wrappers/ZeroExV2ExchangeWrapper.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var web3_utils_1 = __importDefault(require("web3-utils"));
var deployed_json_1 = require("../../migrations/deployed.json");
var ZeroExV2ExchangeWrapper = /** @class */ (function () {
    function ZeroExV2ExchangeWrapper(networkId) {
        this.networkId = networkId;
    }
    ZeroExV2ExchangeWrapper.prototype.getAddress = function () {
        return deployed_json_1.ZeroExV2ExchangeWrapper[this.networkId.toString()].address;
    };
    ZeroExV2ExchangeWrapper.prototype.setNetworkId = function (networkId) {
        this.networkId = networkId;
    };
    ZeroExV2ExchangeWrapper.prototype.orderToBytes = function (order) {
        return []
            .concat(this.toBytes(order.makerAddress))
            .concat(this.toBytes(order.takerAddress))
            .concat(this.toBytes(order.feeRecipientAddress))
            .concat(this.toBytes(order.senderAddress))
            .concat(this.toBytes(order.makerAssetAmount))
            .concat(this.toBytes(order.takerAssetAmount))
            .concat(this.toBytes(order.makerFee))
            .concat(this.toBytes(order.takerFee))
            .concat(this.toBytes(order.expirationTimeSeconds))
            .concat(this.toBytes(order.salt))
            .concat(this.toBytes(order.signature));
    };
    ZeroExV2ExchangeWrapper.prototype.toBytes = function (val) {
        return web3_utils_1.default.hexToBytes(web3_utils_1.default.padLeft(web3_utils_1.default.toHex(val), 64, '0'));
    };
    return ZeroExV2ExchangeWrapper;
}());
exports.ZeroExV2ExchangeWrapper = ZeroExV2ExchangeWrapper;
