// @dydxprotocol/exchange-wrappers/dist/src/types.js
"use strict";
/*

    Copyright 2018 dYdX Trading Inc.

    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.

*/
Object.defineProperty(exports, "__esModule", { value: true });
var OrderType;
(function (OrderType) {
    OrderType["Test"] = "TEST";
    OrderType["ZeroExV2"] = "ZERO_EX_V2";
    OrderType["ZeroExV2MultiOrder"] = "ZERO_EX_V2_MULTI_ORDER";
    OrderType["OasisV1"] = "OASIS_V1";
    OrderType["OasisV2"] = "OASIS_V2";
    OrderType["OasisV3"] = "OASIS_V3";
    OrderType["OasisV3Market"] = "OASIS_V3_MARKET";
    OrderType["OpenDirectly"] = "OPEN_DIRECTLY";
    OrderType["SaiDai"] = "SAI_DAI";
})(OrderType = exports.OrderType || (exports.OrderType = {}));
